app/code/Magento/Catalog/Model/Indexer/Category/Product/AbstractAction.php

```diff
@@ -348,11 +348,7 @@ protected function getNonAnchorCategoriesSelect(Store $store)
                 \Magento\Catalog\Model\Product\Attribute\Source\Status::STATUS_ENABLED
             )->where(
                 $this->connection->getIfNullSql('cpvs.value', 'cpvd.value') . ' IN (?)',
-                [
-                    \Magento\Catalog\Model\Product\Visibility::VISIBILITY_IN_CATALOG,
-                    \Magento\Catalog\Model\Product\Visibility::VISIBILITY_IN_SEARCH,
-                    \Magento\Catalog\Model\Product\Visibility::VISIBILITY_BOTH
-                ]
+                $this->config->getVisibility()->getVisibleInSiteIds()
```

Comment:
I believe that injecting `Visibility` model directly in this class is more correct solution follow [Backward Compatibility](https://developer.adobe.com/commerce/contributor/guides/code-contributions/backward-compatibility-policy/#adding-a-constructor-parameter)

Reply:
Hi @Den4ik